src/alarm_heap.cpp

```diff
@@ -0,0 +1,326 @@
+/**
+* Project Clearwater - IMS in the Cloud
+* Copyright (C) 2016 Metaswitch Networks Ltd
+*
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the
+* Free Software Foundation, either version 3 of the License, or (at your
+* option) any later version, along with the "Special Exception" for use of
+* the program along with SSL, set forth below. This program is distributed
+* in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+* without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+* A PARTICULAR PURPOSE. See the GNU General Public License for more
+* details. You should have received a copy of the GNU General Public
+* License along with this program. If not, see
+* <http://www.gnu.org/licenses/>.
+*
+* The author can be reached by email at clearwater@metaswitch.com or by
+* post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+*
+* Special Exception
+* Metaswitch Networks Ltd grants you permission to copy, modify,
+* propagate, and distribute a work formed by combining OpenSSL with The
+* Software, or a work derivative of such a combination, even if such
+* copying, modification, propagation, or distribution would otherwise
+* violate the terms of the GPL. You must comply with the GPL in all
+* respects for all of the code used other than OpenSSL.
+* "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+* Project and licensed under the OpenSSL Licenses, or a work based on such
+* software and licensed under the OpenSSL Licenses.
+* "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+* under which the OpenSSL Project distributes the OpenSSL toolkit software,
+* as those licenses appear in the file LICENSE-OPENSSL.
+*/
+
+#include <stdexcept>
+#include <time.h>
+#include <cmath>
+#include <arpa/inet.h>
+
+#include "log.h"
+#include "alarm_heap.hpp"
+#include "itu_alarm_table.hpp"
+#include "alarm_active_table.hpp"
+
+AlarmInfo::~AlarmInfo()
+{
+  delete _alarm_in_heap; _alarm_in_heap = NULL;
+}
+
+AlarmSeverityChange AlarmInfo::new_alarm_severity(AlarmDef::Severity severity)
+{
+  // Order the severities so we can do a simple comparison to determine any
+  // severity change.
+  unsigned int ordered_severities[] = {0, 1, 2, 6, 5, 4, 3};
+  int old_severity = ordered_severities[_severity];
+  int new_severity = ordered_severities[severity];
+
+  if (old_severity == new_severity)
+  {
+    return AlarmSeverityChange::NO_CHANGE;
+  }
+  else if (old_severity > new_severity)
+  {
+    return AlarmSeverityChange::REDUCED_SEVERITY;
+  }
+  else
+  {
+    return AlarmSeverityChange::INCREASED_SEVERITY;
+  }
+}
+
+void AlarmInfo::set_severity(AlarmTableDef& alarm_table_def)
```

Comment:
Worth a comment somewhere to say that this class is the one responsible for updating alarmActiveTable behind the scenes?